plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/ControlNodeRule.java

```diff
@@ -0,0 +1,89 @@
+import org.polarsys.capella.transition.system2subsystem.handlers.attachment.FunctionalChainAttachmentHelper;
+import org.polarsys.kitalpha.transposer.rules.handler.rules.api.IContext;
+
+public class ControlNodeRule extends org.polarsys.capella.core.transition.system.rules.fa.ControlNodeRule {
+  @Override
+  public IStatus transformRequired(EObject element, IContext context) {
+    IStatus result = super.transformRequired(element, context);
+    if (result.isOK()) {
+      // A ControlNode is valid if all its source and target FCIFs are valid
+      ControlNode controlNode = (ControlNode) element;
+      FunctionalChainAttachmentHelper helper = FunctionalChainAttachmentHelper.getInstance(context);
+      boolean invalidSource = getSourceFCIFs(controlNode).stream()
+          .anyMatch(inv -> !helper.isValidElement(inv, context));
+      boolean invalidTarget = getTargetFCIFs(controlNode).stream()
+          .anyMatch(inv -> !helper.isValidElement(inv, context));
+      if (invalidSource || invalidTarget) {
+        result = new Status(IStatus.WARNING, Messages.Activity_Transformation,
+            NLS.bind("Control Node ''{0}'' is not valid.", LogHelper.getInstance().getText(controlNode)));
+      }
+    }
+    return result;
+  }
+
+  // TODO: Move me to Capella's code
+  private Set<FunctionalChainInvolvementFunction> getSourceFCIFs(ControlNode controlNode) {
+    Set<FunctionalChainInvolvementFunction> sourceFCIFs = new HashSet<>();
+    List<EObject> incomingSequenceLinks = EObjectExt.getReferencers(controlNode,
+        FaPackage.Literals.SEQUENCE_LINK__TARGET);
+    for (EObject eObj : incomingSequenceLinks) {
+      if (eObj instanceof SequenceLink) {
+        SequenceLink sequenceLink = (SequenceLink) eObj;
+        SequenceLinkEnd source = sequenceLink.getSource();
+        if (source instanceof FunctionalChainInvolvementFunction) {
+          sourceFCIFs.add((FunctionalChainInvolvementFunction) source);
+        } else if (source instanceof ControlNode) {
+          sourceFCIFs.addAll(getSourceFCIFs((ControlNode) source));
+        }
+      }
+    }
+    return sourceFCIFs;
+  }
+
+  // TODO: Move me to Capella's code
+  private Set<FunctionalChainInvolvementFunction> getTargetFCIFs(ControlNode controlNode) {
+    Set<FunctionalChainInvolvementFunction> targetFCIFs = new HashSet<>();
+    List<EObject> outgoingSequenceLinks = EObjectExt.getReferencers(controlNode,
+        FaPackage.Literals.SEQUENCE_LINK__SOURCE);
+    for (EObject eObj : outgoingSequenceLinks) {
+      if (eObj instanceof SequenceLink) {
+        SequenceLink sequenceLink = (SequenceLink) eObj;
+        SequenceLinkEnd target = sequenceLink.getTarget();
+        if (target instanceof FunctionalChainInvolvementFunction) {
+          targetFCIFs.add((FunctionalChainInvolvementFunction) target);
+        } else if (target instanceof ControlNode) {
+          targetFCIFs.addAll(getTargetFCIFs((ControlNode) target));
+        }
+      }
+    }
+    return targetFCIFs;

```

Comment:
Does this work with IT construct ?
![image](https://user-images.githubusercontent.com/34767327/99531157-9d928c80-29a2-11eb-8981-3f733f8ba21f.png)

Reply:
Hum, you're right, it does not work for loop construct. I'll update the code to exclude visited node during the calculation.